File: jobshout-com/web/nextjs/lib/showcase.ts
/**
 * AI Showcase client. Server-side only for anything that carries identity;
 * shares the signed identity headers with Insights.
 */

import type { Job } from "@/lib/api";
import { API_BASE, failure, headers, type Viewer } from "@/lib/insights";

export type Kind = "app" | "agent" | "team";

export const KINDS: Record<Kind, { label: string; plural: string }> = {
  app: { label: "App", plural: "Apps" },
  agent: { label: "Agent", plural: "Agents" },
  team: { label: "Agent team", plural: "Agent teams" },
};

/** Mirrors AGENT_CAPABILITIES in jobshout-domain; the API rejects anything else. */
export const CAPABILITIES = {
  code_generation: "Code generation",
  code_review: "Code review",
  testing: "Testing",
  debugging: "Debugging",
  security: "Security",
  deployment: "Deployment",
  documentation: "Documentation",
  research: "Research",
  data_analysis: "Data analysis",
  design: "Design",
  planning: "Planning",
  operations: "Operations",
} as const;
export type Capability = keyof typeof CAPABILITIES;

export type AppType =
  | "web_application"
  | "mobile_application"
  | "desktop_application"
  | "api"
  | "saas"
  | "ai_application"
  | "agent_application"
  | "mcp_server"
  | "developer_tool"
  | "infrastructure"
  | "security_tool"
  | "data_platform"
  | "rag_application"
  | "automation"
  | "workflow"
  | "game"
  | "marketplace"
  | "other";
export type Maturity =
  | "idea"
  | "prototype"
  | "alpha"
  | "beta"
  | "release_candidate"
  | "production_ready"
  | "enterprise_ready";
export type BuildMethod = "human" | "human_ai" | "ai_assisted" | "agent_built" | "agent_autonomous";
export type Pricing = "open_source" | "free" | "freemium" | "commercial";
export type Visibility = "public" | "unlisted" | "private";
export type AppStatus = "draft" | "pending_review" | "published" | "rejected" | "archived";
export type Verification =
  | "unverified"
  | "creator_verified"
  | "repository_verified"
  | "build_verified"
  | "security_scanned"
  | "production_verified";
export type Collection = "featured" | "production_ready" | "built_by_agents" | "open_source" | "hiring";
export type SortKey = "new" | "stars" | "updated";

export const APP_TYPES: Record<AppType, string> = {
  web_application: "Web app",
  mobile_application: "Mobile app",
  desktop_application: "Desktop app",
  api: "API",
  saas: "SaaS",
  ai_application: "AI app",
  agent_application: "Agent app",
  mcp_server: "MCP server",
  developer_tool: "Developer tool",
  infrastructure: "Infrastructure",
  security_tool: "Security tool",
  data_platform: "Data platform",
  rag_application: "RAG app",
  automation: "Automation",
  workflow: "Workflow",
  game: "Game",
  marketplace: "Marketplace",
  other: "Other",
};

export const MATURITY: Record<Maturity, { label: string; blurb: string }> = {
  idea: { label: "Idea", blurb: "A concept, maybe a sketch." },
  prototype: { label: "Prototype", blurb: "Works on the happy path." },
  alpha: { label: "Alpha", blurb: "Early users, rough edges expected." },
  beta: { label: "Beta", blurb: "Feature-complete, still hardening." },
  release_candidate: { label: "Release candidate", blurb: "Ready unless something turns up." },
  production_ready: {
    label: "Production ready",
    blurb: "Tested, monitored and documented, with a live deployment.",
  },
  enterprise_ready: {
    label: "Enterprise ready",
    blurb: "Production ready, plus dependency scanning, backups and a release history.",
  },
};

export const BUILD_METHODS: Record<BuildMethod, { label: string; blurb: string }> = {
  human: { label: "Built by humans", blurb: "No AI in the build." },
  ai_assisted: { label: "AI assisted", blurb: "People wrote it, with AI help on parts." },
  human_ai: { label: "Human + AI", blurb: "People and AI built it together." },
  agent_built: { label: "Agent built", blurb: "Agents did most of the building; people steered." },
  agent_autonomous: {
    label: "Agent autonomous",
    blurb: "Agents built it end to end; people approved.",
  },
};

export const PRICING: Record<Pricing, string> = {
  open_source: "Open source",
  free: "Free",
  freemium: "Freemium",
  commercial: "Commercial",
};

export const VISIBILITY: Record<Visibility, { label: string; blurb: string }> = {
  public: { label: "Public", blurb: "Listed in the showcase and search once approved." },
  unlisted: { label: "Unlisted", blurb: "Anyone with the link, kept out of lists and search." },
  private: { label: "Private", blurb: "Only you and the editors." },
};

export const STATUS_LABELS: Record<AppStatus, string> = {
  draft: "Draft",
  pending_review: "In review",
  published: "Published",
  rejected: "Changes requested",
  archived: "Archived",
};

export const VERIFICATION_LABELS: Record<Verification, string> = {
  unverified: "Unverified",
  creator_verified: "Creator verified",
  repository_verified: "Repository verified",
  build_verified: "Build verified",
  security_scanned: "Security scanned",
  production_verified: "Production verified",
};

export const COLLECTIONS: Record<Collection, { label: string; blurb: string }> = {
  featured: { label: "Featured", blurb: "Picked by the JobShout editors." },
  production_ready: {
    label: "Production ready",
    blurb: "Creators who back the claim with tests, CI, monitoring and a live deployment.",
  },
  built_by_agents: {
    label: "Built by agents",
    blurb: "Apps where AI agents did most of the building.",
  },
  open_source: { label: "Open source", blurb: "Read the code, run it yourself." },
  hiring: { label: "Hiring", blurb: "Projects with open roles on the JobShout board." },
};

export const EVIDENCE: Array<{ key: EvidenceFlag; label: string }> = [
  { key: "automated_tests", label: "Automated tests" },
  { key: "ci_cd", label: "CI/CD" },
  { key: "security_scanning", label: "Security scanning" },
  { key: "dependency_scanning", label: "Dependency scanning" },
  { key: "monitoring", label: "Monitoring" },
  { key: "backups", label: "Backups and recovery" },
  { key: "documentation", label: "Documentation" },
  { key: "release_history", label: "Release history" },
];

export type EvidenceFlag =
  | "automated_tests"
  | "ci_cd"
  | "security_scanning"
  | "dependency_scanning"
  | "monitoring"
  | "backups"
  | "documentation"
  | "release_history";

export type Evidence = Record<EvidenceFlag, boolean> & { status_page_url: string };

export interface ShowcaseAgent {
  name: string;
  role: string;
}

/** A directory agent or team another entry links to. */
export interface ShowcaseLink {
  slug: string;
  kind: Kind;
  name: string;
  tagline: string;
  logo_url: string;
  role: string;
}

export interface LinkInput {
  slug: string;
  role: string;
}

export interface ShowcaseApp {
  id: string;
  kind: Kind;
  slug: string;
  name: string;
  tagline: string;
  description_md: string;
  description_html: string;
  app_type: AppType;
  maturity: Maturity;
  build_method: BuildMethod;
  pricing: Pricing;
  license: string;
  version: string;
  logo_url: string;
  screenshots: string[];
  repo_url: string;
  demo_url: string;
  website_url: string;
  docs_url: string;
  technologies: string[];
  ai_models: string[];
  agents: ShowcaseAgent[];
  human_oversight: string;
  evidence: Evidence;
  team_name: string;
  model_provider: string;
  tools: string[];
  mcp_servers: string[];
  capabilities: string[];
  linked_agents: ShowcaseLink[];
  linked_team: ShowcaseLink | null;
  used_in: number;
  /** Open roles on the board this entry is hiring for (published jobs only). */
  jobs: Job[];
  creator_email?: string;
  creator_display_name: string;
  visibility: Visibility;
  status: AppStatus;
  verification: Verification;
  featured: boolean;
  review_note: string;
  star_count: number;
  starred: boolean;
  published_at?: string | null;
  created_at: string;
  updated_at: string;
}

export type ShowcaseAppInput = Omit<
  ShowcaseApp,
  | "id"
  | "slug"
  | "description_html"
  | "creator_email"
  | "creator_display_name"
  | "status"
  | "verification"
  | "featured"
  | "review_note"
  | "star_count"
  | "starred"
  | "published_at"
  | "created_at"
  | "updated_at"
  | "linked_agents"
  | "linked_team"
  | "used_in"
  | "jobs"
> & { submit: boolean; agent_links: LinkInput[]; team_slug: string; job_ids: string[] };

export interface ShowcaseTag {
  name: string;
  count: number;
}

export type ShowcaseQuery = {
  kind?: Kind | null;
  capability?: Capability | null;
  q?: string | null;
  type?: AppType | null;
  collection?: Collection | null;
  tech?: string | null;
  sort?: SortKey | null;
  limit?: number;
  offset?: number;
};

const is =
  <T extends string>(values: Record<T, unknown>) =>
  (v: unknown): v is T =>
    typeof v === "string" && Object.prototype.hasOwnProperty.call(values, v);

export const isAppType = is(APP_TYPES);
export const isMaturity = is(MATURITY);
export const isBuildMethod = is(BUILD_METHODS);
export const isPricing = is(PRICING);
export const isVisibility = is(VISIBILITY);
export const isCollection = is(COLLECTIONS);
export const isKind = is(KINDS);
export const isCapability = is(CAPABILITIES);
export const isSortKey = (v: unknown): v is SortKey => v === "new" || v === "stars" || v === "updated";

export async function listApps(
  query: ShowcaseQuery = {},
  viewer?: Viewer | null,
): Promise<{ data: ShowcaseApp[]; total: number }> {
  const params = new URLSearchParams();
  if (query.kind) params.set("kind", query.kind);
  if (query.capability) params.set("capability", query.capability);
  if (query.q) params.set("q", query.q);
  if (query.type) params.set("type", query.type);
  if (query.collection) params.set("collection", query.collection);
  if (query.tech) params.set("tech", query.tech);
  if (query.sort) params.set("sort", query.sort);
  params.set("limit", String(query.limit ?? 12));
  if (query.offset) params.set("offset", String(query.offset));
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps?${params}`, {
    headers: headers(viewer),
    // Per-viewer star state makes signed-in lists uncacheable.
    ...(viewer ? { cache: "no-store" as const } : { next: { revalidate: 30, tags: ["showcase"] } }),
  });
  if (!res.ok) throw await failure(res, "Failed to load the showcase");
  return (await res.json()) as { data: ShowcaseApp[]; total: number };
}

export async function listTags(limit = 24, kind: Kind = "app"): Promise<ShowcaseTag[]> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/tags?limit=${limit}&kind=${kind}`, {
    next: { revalidate: 300, tags: ["showcase"] },
  });
  if (!res.ok) throw await failure(res, "Failed to load technologies");
  return ((await res.json()) as { data: ShowcaseTag[] }).data;
}

export async function getApp(slug: string, viewer?: Viewer | null): Promise<ShowcaseApp | null> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps/${encodeURIComponent(slug)}`, {
    headers: headers(viewer),
    cache: "no-store",
  });
  if (res.status === 404) return null;
  if (!res.ok) throw await failure(res, "Failed to load the app");
  return (await res.json()) as ShowcaseApp;
}

export async function relatedApps(slug: string): Promise<ShowcaseApp[]> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps/${encodeURIComponent(slug)}/related`, {
    next: { revalidate: 60, tags: ["showcase"] },
  });
  if (!res.ok) return [];
  return ((await res.json()) as { data: ShowcaseApp[] }).data;
}

/** Public apps (and, for an agent, teams) that link to an agent or team. */
export async function usedIn(
  slug: string,
  viewer?: Viewer | null,
): Promise<{ apps: ShowcaseApp[]; teams: ShowcaseApp[] }> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps/${encodeURIComponent(slug)}/used-in`, {
    headers: headers(viewer),
    cache: "no-store",
  });
  if (!res.ok) return { apps: [], teams: [] };
  return (await res.json()) as { apps: ShowcaseApp[]; teams: ShowcaseApp[] };
}

/** Agents or teams the viewer may link: public ones plus their own. */
export async function linkCandidates(
  viewer: Viewer,
  kind: "agent" | "team",
  q: string,
): Promise<ShowcaseApp[]> {
  const params = new URLSearchParams({ kind });
  if (q.trim()) params.set("q", q.trim());
  const res = await fetch(`${API_BASE}/api/v1/showcase/link-candidates?${params}`, {
    headers: headers(viewer),
    cache: "no-store",
  });
  if (!res.ok) throw await failure(res, "Could not search the directory");
  return ((await res.json()) as { data: ShowcaseApp[] }).data;
}

/** Open jobs the viewer may link: the ones they posted (editors: any). */
export async function linkableJobs(viewer: Viewer): Promise<Job[]> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/linkable-jobs`, {
    headers: headers(viewer),
    cache: "no-store",
  });
  if (!res.ok) return [];
  return ((await res.json()) as { data: Job[] }).data;
}

/** Public showcase entries (any kind) hiring for this job. */
export async function entriesForJob(jobId: string): Promise<ShowcaseApp[]> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/by-job/${encodeURIComponent(jobId)}`, {
    next: { revalidate: 60, tags: ["showcase"] },
  });
  if (!res.ok) return [];
  return ((await res.json()) as { data: ShowcaseApp[] }).data;
}

export async function myApps(viewer: Viewer): Promise<ShowcaseApp[]> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps/mine`, {
    headers: headers(viewer),
    cache: "no-store",
  });
  if (!res.ok) throw await failure(res, "Failed to load your apps");
  return ((await res.json()) as { data: ShowcaseApp[] }).data;
}

/** Returns null when the viewer is not an editor. */
export async function showcaseReviewQueue(viewer: Viewer): Promise<ShowcaseApp[] | null> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/review-queue`, {
    headers: headers(viewer),
    cache: "no-store",
  });
  if (res.status === 403) return null;
  if (!res.ok) throw await failure(res, "Failed to load the review queue");
  return ((await res.json()) as { data: ShowcaseApp[] }).data;
}

export async function saveApp(
  viewer: Viewer,
  input: ShowcaseAppInput,
  id?: string,
): Promise<ShowcaseApp> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps${id ? `/${id}` : ""}`, {
    method: id ? "PATCH" : "POST",
    headers: headers(viewer),
    body: JSON.stringify(input),
    cache: "no-store",
  });
  if (!res.ok) throw await failure(res, "Failed to save");
  return (await res.json()) as ShowcaseApp;
}

export async function deleteApp(viewer: Viewer, id: string): Promise<void> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps/${id}`, {
    method: "DELETE",
    headers: headers(viewer),
    cache: "no-store",
  });
  if (!res.ok) throw await failure(res, "Failed to delete");
}

export type ShowcaseModeration = "approve" | "reject" | "feature" | "unfeature" | "archive";

export async function moderateApp(
  viewer: Viewer,
  id: string,
  action: ShowcaseModeration,
  note = "",
): Promise<ShowcaseApp> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps/${id}/moderate`, {
    method: "POST",
    headers: headers(viewer),
    body: JSON.stringify({ action, note }),
    cache: "no-store",
  });
  if (!res.ok) throw await failure(res, "Failed to update");
  return (await res.json()) as ShowcaseApp;
}

export async function setStar(
  viewer: Viewer,
  slug: string,
  on: boolean,
): Promise<{ starred: boolean; star_count: number }> {
  const res = await fetch(`${API_BASE}/api/v1/showcase/apps/${encodeURIComponent(slug)}/star`, {
    method: on ? "PUT" : "DELETE",
    headers: headers(viewer),
    cache: "no-store",
  });
  if (!res.ok) throw await failure(res, "Failed to update the star");
  return (await res.json()) as { starred: boolean; star_count: number };
}

/** Where an entry lives: apps in the showcase, agents and teams in the directory. */
export function entryHref(e: { kind: Kind; slug: string }): string {
  return e.kind === "app" ? `/showcase/${e.slug}` : `/agents/${e.slug}`;
}

export function directoryHref(f: { kind?: Kind | null; capability?: Capability | null; q?: string | null; tech?: string | null; sort?: SortKey | null; page?: number }): string {
  const p = new URLSearchParams();
  if (f.kind === "team") p.set("kind", "team");
  if (f.capability) p.set("capability", f.capability);
  if (f.tech) p.set("tech", f.tech);
  if (f.q) p.set("q", f.q);
  if (f.sort && f.sort !== "new") p.set("sort", f.sort);
  if (f.page && f.page > 1) p.set("page", String(f.page));
  const s = p.toString();
  return s ? `/agents?${s}` : "/agents";
}

/** "github.com/acme/app" — a link's host and path, for display. */
export function displayUrl(url: string): string {
  return url.replace(/^https?:\/\/(www\.)?/, "").replace(/\/$/, "");
}

export function showcaseHref(f: ShowcaseQuery & { page?: number }): string {
  const p = new URLSearchParams();
  if (f.collection) p.set("collection", f.collection);
  if (f.type) p.set("type", f.type);
  if (f.tech) p.set("tech", f.tech);
  if (f.q) p.set("q", f.q);
  if (f.sort && f.sort !== "new") p.set("sort", f.sort);
  if (f.page && f.page > 1) p.set("page", String(f.page));
  const s = p.toString();
  return s ? `/showcase?${s}` : "/showcase";
}

File: jobshout-com/crates/jobshout-domain/src/showcase.rs
//! AI Showcase: applications built by people and AI agents.
//!
//! Everything a creator says about an app (maturity, how it was built, its
//! production evidence) is self-declared. `verification` is the only field
//! that says what JobShout itself has checked, and nothing sets it above
//! `unverified` yet.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

pub type ShowcaseAppId = Uuid;

string_enum! {
    /// What a showcase entry is. One table, like Insights: every kind shares
    /// review, stars, search, visibility and "Your apps".
    ShowcaseKind {
        App => "app",
        Agent => "agent",
        Team => "team",
    }
}

/// What an agent can do, as the directory filters by it. Stored as strings.
pub const AGENT_CAPABILITIES: &[&str] = &[
    "code_generation",
    "code_review",
    "testing",
    "debugging",
    "security",
    "deployment",
    "documentation",
    "research",
    "data_analysis",
    "design",
    "planning",
    "operations",
];

string_enum! {
    ShowcaseAppType {
        WebApplication => "web_application",
        MobileApplication => "mobile_application",
        DesktopApplication => "desktop_application",
        Api => "api",
        Saas => "saas",
        AiApplication => "ai_application",
        AgentApplication => "agent_application",
        McpServer => "mcp_server",
        DeveloperTool => "developer_tool",
        Infrastructure => "infrastructure",
        SecurityTool => "security_tool",
        DataPlatform => "data_platform",
        RagApplication => "rag_application",
        Automation => "automation",
        Workflow => "workflow",
        Game => "game",
        Marketplace => "marketplace",
        Other => "other",
    }
}

string_enum! {
    /// How far along the app is, in the creator's words.
    ShowcaseMaturity {
        Idea => "idea",
        Prototype => "prototype",
        Alpha => "alpha",
        Beta => "beta",
        ReleaseCandidate => "release_candidate",
        ProductionReady => "production_ready",
        EnterpriseReady => "enterprise_ready",
    }
}

impl ShowcaseMaturity {
    /// The two levels that must be backed by production evidence.
    pub fn needs_evidence(self) -> bool {
        matches!(self, Self::ProductionReady | Self::EnterpriseReady)
    }
}

string_enum! {
    /// Who did the building.
    ShowcaseBuildMethod {
        Human => "human",
        HumanAi => "human_ai",
        AiAssisted => "ai_assisted",
        AgentBuilt => "agent_built",
        AgentAutonomous => "agent_autonomous",
    }
}

impl ShowcaseBuildMethod {
    pub fn is_agent_built(self) -> bool {
        matches!(self, Self::AgentBuilt | Self::AgentAutonomous)
    }
}

string_enum! {
    ShowcasePricing {
        OpenSource => "open_source",
        Free => "free",
        Freemium => "freemium",
        Commercial => "commercial",
    }
}

string_enum! {
    /// Private: owner and editors only. Unlisted: anyone with the link once
    /// published, but kept out of lists and search. Public: everywhere.
    ShowcaseVisibility {
        Public => "public",
        Unlisted => "unlisted",
        Private => "private",
    }
}

string_enum! {
    ShowcaseStatus {
        Draft => "draft",
        PendingReview => "pending_review",
        Published => "published",
        Rejected => "rejected",
        Archived => "archived",
    }
}

string_enum! {
    /// What JobShout has checked. Only `unverified` is reachable today; the
    /// higher levels arrive with the verification pipeline.
    ShowcaseVerification {
        Unverified => "unverified",
        CreatorVerified => "creator_verified",
        RepositoryVerified => "repository_verified",
        BuildVerified => "build_verified",
        SecurityScanned => "security_scanned",
        ProductionVerified => "production_verified",
    }
}

/// An agent that worked on the app, as the creator describes it.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ShowcaseAgent {
    pub name: String,
    #[serde(default)]
    pub role: String,
}

/// What the creator says backs a production claim. Self-declared.
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProductionEvidence {
    #[serde(default)]
    pub automated_tests: bool,
    #[serde(default)]
    pub ci_cd: bool,
    #[serde(default)]
    pub security_scanning: bool,
    #[serde(default)]
    pub dependency_scanning: bool,
    #[serde(default)]
    pub monitoring: bool,
    #[serde(default)]
    pub backups: bool,
    #[serde(default)]
    pub documentation: bool,
    #[serde(default)]
    pub release_history: bool,
    /// Public status or uptime page, if there is one.
    #[serde(default)]
    pub status_page_url: String,
}

/// Another showcase entry this one links to: an agent that built an app, the
/// team that did, or a team's member. Only entries the viewer may see appear.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ShowcaseLink {
    pub slug: String,
    pub kind: ShowcaseKind,
    pub name: String,
    pub tagline: String,
    pub logo_url: String,
    /// What it did here ("Backend", "Reviewed the auth flow").
    pub role: String,
}

/// A link as a creator submits it: the target's slug and its role.
#[derive(Debug, Clone, Default, PartialEq, Eq, Deserialize)]
pub struct ShowcaseLinkInput {
    pub slug: String,
    #[serde(default)]
    pub role: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShowcaseApp {
    pub id: ShowcaseAppId,
    pub kind: ShowcaseKind,
    pub slug: String,
    pub name: String,
    pub tagline: String,
    pub description_md: String,
    /// Sanitised server-side from description_md; the web renders this only.
    pub description_html: String,
    pub app_type: ShowcaseAppType,
    pub maturity: ShowcaseMaturity,
    pub build_method: ShowcaseBuildMethod,
    pub pricing: ShowcasePricing,
    pub license: String,
    pub version: String,
    pub logo_url: String,
    pub screenshots: Vec<String>,
    pub repo_url: String,
    pub demo_url: String,
    pub website_url: String,
    pub docs_url: String,
    pub technologies: Vec<String>,
    pub ai_models: Vec<String>,
    /// Agents named in free text (not in the directory).
    pub agents: Vec<ShowcaseAgent>,
    /// Where humans reviewed or approved the agents' work.
    pub human_oversight: String,
    pub evidence: ProductionEvidence,
    pub team_name: String,
    /// Agents only: who serves the model(s) in `ai_models`.
    pub model_provider: String,
    /// Agents only: tools it can call.
    pub tools: Vec<String>,
    /// Agents only: MCP servers it uses.
    pub mcp_servers: Vec<String>,
    /// Agents only: values from [`AGENT_CAPABILITIES`].
    pub capabilities: Vec<String>,
    /// Apps: directory agents that built it. Teams: members, in workflow order.
    pub linked_agents: Vec<ShowcaseLink>,
    /// Apps only: the directory team that built it.
    pub linked_team: Option<ShowcaseLink>,
    /// Agents and teams: public, published apps that link to it (directly,
    /// or through a team for agents).
    pub used_in: i64,
    /// Only returned to the creator and editors; public responses blank it.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub creator_email: Option<String>,
    pub creator_display_name: String,
    pub visibility: ShowcaseVisibility,
    pub status: ShowcaseStatus,
    pub verification: ShowcaseVerification,
    pub featured: bool,
    pub review_note: String,
    pub star_count: i32,
    /// Whether the viewer has starred it; false for anonymous viewers.
    pub starred: bool,
    pub published_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Create and update share one body: an edit sends the whole app again.
#[derive(Debug, Clone, Default, Deserialize)]
pub struct ShowcaseAppInput {
    /// Required on create; an entry never changes kind.
    pub kind: Option<ShowcaseKind>,
    #[serde(default)]
    pub name: String,
    #[serde(default)]
    pub tagline: String,
    #[serde(default)]
    pub description_md: String,
    pub app_type: Option<ShowcaseAppType>,
    pub maturity: Option<ShowcaseMaturity>,
    pub build_method: Option<ShowcaseBuildMethod>,
    pub pricing: Option<ShowcasePricing>,
    pub visibility: Option<ShowcaseVisibility>,
    #[serde(default)]
    pub license: String,
    #[serde(default)]
    pub version: String,
    #[serde(default)]
    pub logo_url: String,
    #[serde(default)]
    pub screenshots: Vec<String>,
    #[serde(default)]
    pub repo_url: String,
    #[serde(default)]
    pub demo_url: String,
    #[serde(default)]
    pub website_url: String,
    #[serde(default)]
    pub docs_url: String,
    #[serde(default)]
    pub technologies: Vec<String>,
    #[serde(default)]
    pub ai_models: Vec<String>,
    #[serde(default)]
    pub agents: Vec<ShowcaseAgent>,
    #[serde(default)]
    pub human_oversight: String,
    #[serde(default)]
    pub evidence: ProductionEvidence,
    #[serde(default)]
    pub team_name: String,
    #[serde(default)]
    pub model_provider: String,
    #[serde(default)]
    pub tools: Vec<String>,
    #[serde(default)]
    pub mcp_servers: Vec<String>,
    #[serde(default)]
    pub capabilities: Vec<String>,
    /// Apps: directory agents that built it. Teams: members, in order.
    #[serde(default)]
    pub agent_links: Vec<ShowcaseLinkInput>,
    /// Apps only: slug of the directory team that built it, or empty.
    #[serde(default)]
    pub team_slug: String,
    /// false saves a draft; true submits (community) or publishes (editors).
    #[serde(default)]
    pub submit: bool,
}

/// A technology tag and how many public apps use it.
#[derive(Debug, Clone, Serialize)]
pub struct ShowcaseTag {
    pub name: String,
    pub count: i64,
}
